src/Microsoft.IdentityModel.Logging/LogHelper.cs

```diff
@@ -358,9 +360,42 @@ public static string FormatInvariant(string format, params object[] args)
                 return format;
 
             if (!IdentityModelEventSource.ShowPII)
-                return string.Format(CultureInfo.InvariantCulture, format, args.Select(RemovePII).ToArray()); 
+                return string.Format(CultureInfo.InvariantCulture, format, args.Select(RemovePII).ToArray());
+            else
+                return string.Format(CultureInfo.InvariantCulture, format, args.Select(SanitizeSecurityArtifact).ToArray());
+        }
+
+        /// <summary>
+        /// Regex that is used to figure out if a token is in JWS format.
+        /// </summary>
+        public static readonly Regex RegexJws = new Regex(JsonCompactSerializationRegex, RegexOptions.Compiled | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
+
+        /// <summary>
+        /// Regex that is used to figure out if a token is in JWE format.
+        /// </summary>
+        public static readonly Regex RegexJwe = new Regex(JweCompactSerializationRegex, RegexOptions.Compiled | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
+
+        /// <summary>
+        /// JWS - Token format: 'header.payload.signature'. Signature is optional, but '.' is required.
+        /// </summary>
+        public const string JsonCompactSerializationRegex = @"^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*$";
 
-            return string.Format(CultureInfo.InvariantCulture, format, args);
+        /// <summary>
+        /// JWE - Token format: 'protectedheader.encryptedkey.iv.cyphertext.authenticationtag'.
+        /// </summary>
+        public const string JweCompactSerializationRegex = @"^[A-Za-z0-9-_]+\.[A-Za-z0-9-_]*\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+\.[A-Za-z0-9-_]+$";
+
+        private static string SanitizeSecurityArtifact(object arg)
+        {
+            if (arg is ISafeLogSecurityArtifact)
```

Comment:
>ISafeLogSecurityArtifact [](http://example.com/codeflow?start=23&length=24)

If you are planning to use the ISafeLogSecurityArtifact interface, there is no need for the SecurityArtifact type and the related helper methods to wrap objects with this type.

Reply:
method could have the argument be of type SecurityArtifact instead. right now you are typecasting 3 places. and every path has 2 type casts

Reply:
Several places in our call graphs log tokens as strings, since they are not ISafeLogSecurityArtifact, ToString() will be called and the entire token with signature will be logged.